src/Type/Schema.php

```diff
@@ -57,7 +57,7 @@ class Schema
      */
     private $resolvedTypes = [];
 
-    /** @var array<string, array<string, ObjectType>>|null */
+    /** @var array<string, array<string, ObjectType|UnionType>> */
```

Comment:
This seems it can still be null

Reply:
Also the UnionType is missing in the getter method return type

Reply:
> This seems it can still be null

Where do you see it receiving null?

> Also the UnionType is missing in the getter method return type

👀 e73353c

Reply:
```php
    public function testNull()
    {
        $schema = new Schema([]);
        self::assertNotNull($schema->possibleTypeMap);
    }
```

this fails because `Schema::$possibleTypeMap` can be null but it's not listed in `@var` annotation.

Reply:
`possibleTypeMap` is private, is it not?

Reply:
Yes. This was just an example to show the issue.

Reply:
But it doesn't show the issue. It's private, so your assert wouldn't even get a chance to fire.

Reply:
But it's still null for class members after instantiation. So there's no way to eg. statically analyses that the value is set. It's bug of this kind https://3v4l.org/jMf4f

Reply:
> But it's still null for class members after instantiation. 

By that logic, literally all class members that are not initialized at declaration time should have `null` as part of their type, and I don't think you want that. I should also mention that as of PHP 7.4, if you try to access a typehinted member before it has been initialized you will get an exception of the form "must not be accessed before initializtion"; in other words, the default value is not `null`, it is simply undefined.

Reply:
I think lazily initialized properties such as this one should be annotated as nullable, consider https://github.com/webonyx/graphql-php/blob/f6839dd6559bc0f891fdb1af1e1bc9c00ad32aae/src/Type/Schema.php#L409

When we switch to PHP 7.4 and use typed properties, we will have to choose one of those variants:

```php
private int $y;

if (! isset($this->y))
```

```php
private ?int $y;

if ($this->y === null)
```

I prefer the first one, because it disallows `$this->y = null`. So to summarize, i think the type annotation is good, and we might switch the check in the lazy access to use `isset()`.

Reply:
> So to summarize, i think the type annotation is good, and we might switch the check in the lazy access to use isset().

Are you arguing for or against adding the `null` type? A shift towards `isset` seems to be in favor of not using it...

Reply:
Non-null, as it sits in your PR 👍

Reply:
> By that logic, literally all class members that are not initialised at declaration time should have null as part of their type

exactly

I created wrong example before, this is the bug we're introducing here https://3v4l.org/5Fvv3

We have not yet dropped php 7.1, we're not dropping 7.3 anytime soon that said for following few years there won't be typed properties in this lib.

Also any type checker will complain about this code `if ($this->possibleTypeMap === null) {`  because we're wrongly introducing statement that it can never be `null`. I must admit that I really can't understand while we all know the `null` is property's possible type why we're discussing whether we document it or not. Especially considering the fact it was correct before.

Reply:
That conflates *declaration* with *assignment*. We should change the check to `if (! isset($this->possibleTypeMap)) {`.

There is value in declaring properties as non-null: Assigning `null` is not allowed. We already have a way of knowing if a property is always initialized (if it is assigned a value at declaration).

Those semantics don't really change from PHP 7.1 to PHP 7.4, they are just reiffied by the actual property types.

> By that logic, literally all class members that are not initialised at declaration time should have null as part of their type

Reply:
But this is not about the actual check. That's about introducing bug within the class scope.

1) Schema can be instantiated with invalid state.
2) Whenever working within the class with `$this->possibleTypeMap` there's higher risk of breaking something because `null` declaration is missing so we can't leverage type analysis to catch accessing the empty value as shown in my 3v4l.
3) This https://3v4l.org/Ospoj is ofc nice but we will won't have typed properties any time in the near future so there's no point writing the code like we have them already.

We will have this for few years
```php
    /** @var int */
    private $y;
```

Not this

```php
    /** @var int|null */
    private $y;
```

Therefore this is unsafe and bug https://phpstan.org/r/21100d5a-c0d0-4e74-8eba-884961ab7d29 while this is still bug https://phpstan.org/r/58a50994-2ba0-405c-8ed5-5c7c57bd5470 but safely caught.

When we drop PHP 7.3 support, we can start working with typed properties.

Reply:
> we can't leverage type analysis to catch accessing the empty value

That is a limitation of PHPStan, see https://github.com/phpstan/phpstan-strict-rules/issues/88

It is unfortunate the tools don't support it, but the semantics of this definition are quite clear to me:

```php
    /** @var int */
    private $y;
```

It is obvious that `$y` isn't an `int` all the time, but the definition tells me *if* it is initialized, it may only be assigned to an `int`.

Consider the following bug: https://3v4l.org/PoOhI

Reply:
> b02d8ce

Actually, I'm a little baffled. I have the commit locally and pushed to this branch, but git is telling me that "Everything up-to-date", and I don't see it here. 😕

Reply:
BTW I think this is not valid even with php 7.4+. If property is null after construction, it should be marked as nullable. We won't be able to fully leverage static analysis over this if we introduce this bug.

Reply:
It is not `null` though, it is uninitialized. It just so happens that untyped properties return `null` when accessed, whereas typed properties throw.

Reply:
Can you share such a snippet that is valid with psalm?

Reply:
> We won't be able to fully leverage static analysis over this if we introduce this bug.

What bug? Semantically, Uninitialized variables are not `null`, they are undefined. Yes, for the moment, in PHP they take the `null` value, but this is an implementation detail that can change (as we will see in PHP 7.4). It doesn't matter what happens in the constructor, it doesn't matter when they get their first write, and it doesn't even matter that they are members (the same principle holds for locally declared variables); all that matters is that you should not try to access it until it has been initialized. True, PHPStan does not currently catch usage of uninitialized variables, but that's a limitation, not a feature.

I feel like we're going in circles, here. What is it going to take to convince you?

Reply:
The thing is that it is a feature https://github.com/vimeo/psalm/issues/2650

https://psalm.dev/articles/psalm-supports-php-7-4#warn-about-properties-not-initialised-within-constructors

Reply:
I don't understand; what impact does that have on this discussion? The sample you shared doesn't seem to have anything to do with nullable types.

Reply:
We're doing this https://psalm.dev/r/5c9d1dcdd0. This will always emit errors.

Reply:
This is _enforcement_ of not accessing uninitialized varaibles. It's great, and I wish PHPStan did it, too (no doubt it will, someday), but it does not change or conflict with what @spawnia and I are trying to tell you.

Reply:
I guess I still don't see how do we make the code analysable without ignoring such errors while maintaining this suggested style of accessing variables.

Reply:
We don't currently have any errors. We are correctly using `isset` to ensure that we don't access a member that we know can be undefined. The code is happy. The functionality is happy. The static checker is happy. The only small wrinkle is that we're missing a nice-to-have, which is the ability for our checker to catch usage of uninitialized variables, but even if the checker never gets around to it, it will happen natively for us all in PHP 7.4.

Reply:
So as we target to max level of static analysis, we will have to ignore errors reported by code analysis in the future to bypass code style like this. And that will result in rendering the code less error prone because of unnecessary ignores in our checks.

I'm not saying it does not work. I'm saying this will lower our code quality as we will have to disable our future checks. Static checker is happy because most of it's functionality is disabled...

Reply:
> we will have to ignore errors reported by code analysis in the future to bypass code style like this

No. The checker will not produce any errors here, even at max level. The worst you can say about it is that it will fail to catch usage of uninitialized variables. That's regrettable, but it will either appear in time, or eventually be mooted by PHP 7.4. It's not just cause to run around through our code making all of our members nullable, which absolutely WILL weaken our code (by inviting null writes that are not appropriate).

I can't spend the rest of my life arguing about this. Take one more day to sleep on it, and if you're still not convinced I'll change this one instance just to get us moving again, and we'll resume the fight some other day.

Reply:
Psalm won't even allow nullable typed properties to be uninitialized after construction: https://psalm.dev/r/a982c0ab46. It's not like this approach is foolproof: https://psalm.dev/r/3342e252f2 fails https://3v4l.org/jBLFS

Lazy property initialization is useful for performance. I also want to be able to use typed properties once they are available, and to write idiomatic code. Changing the type hint to allow something wrong (assigning `null`) to work around the limitations of tools we are using seems silly.

A great benefit of non-null types is that once we know the property was initialized, we know it's value is good. When we make every property nullable, our code is going to be littered with null-checks.

Hopefully, the tooling will grow to catch more real errors related to possibly uninitialized properties. Until then, we can ignore rules that don't help us to achieve the result we want. Luckily for us, it's not like there is some kind of type police going around that is going to beat us for it 👮‍♀️

Reply:
> When we make every property nullable, our code is going to be littered with null-checks.

Thank you 👍

Reply:
> Psalm won't even allow nullable typed properties to be uninitialized after construction

That strikes me as a kind of hammy check--there is no requirement in any other statically typed language I know of to initialize your members in the constructor. Who says that's appropriate in every case? I suspect they did it this way because it was simpler than tracking the state of the variable through the entire tree during analysis.

Reply:
So as I said it basically means we will have to suppress errors from our static analysis tooling. I tried to avoid that to keep code quality automated as much as possible.

Reply:
> So as I said it basically means we will have to suppress errors from our static analysis tooling. 

What errors?

Reply:
I linked it before, this error case

https://psalm.dev/r/5c9d1dcdd0  
https://github.com/vimeo/psalm/issues/2650

Reply:
Yes, that is psalm. They took a very conservative approach to checking for uninitialized members (the hammy constructor-time check we've already discussed). PHPStan doesn't do it that way, and it never will. We don't currently have any static checking errors around our current approach, and we have no reason to expect any in the future.